### PR TITLE
Let's fix instead of reverting flambda_backend_args

### PR DESCRIPTION
**Pull request by gretay-js**
Revert #395 and fix #382.

### CODE REVIEW

**Comment by gretay-js** [comment]
@lukemaurer I added some comments. In the process, I've noticed that the name of the module and module type are confusingly the same, so I changed it.

File: driver/flambda_backend_args.mli [hunk 1]
```diff
@@ -13,6 +13,23 @@
 (*   special exception on linking described in the file LICENSE.          *)
 (*                                                                        *)
 (**************************************************************************)
+(** This module follows the structure of ocaml/driver/main_args.ml{i}.
+    It provides a way to (a) share argument implementations between
+    different installable tools and (b) override default implementations
+    of arguments. *)
+
+(** Steps for adding a new flag:
```
Comment (gretay-js) [inline driver/flambda_backend_args.mli:21]: The "steps" could go into HACKING doc.

**Comment by mshinwell** [comment]
From talking to @lukemaurer I understand it might be the case that the underscores in the variable names in our new file might have been the reason we didn't catch the bug in the previous version of this patch.  Is that correct?  If so let's please remove all of the underscores.  (Even if not, maybe we should remove them anyway.)

**Comment by mshinwell** [comment]
I was just talking about `Flambda_backend_options`, not the `ocaml` subdirectory.

File: driver/flambda_backend_args.mli (hunk 1 above)
Comment (mshinwell) [reply driver/flambda_backend_args.mli:21]: Yes, please move them.

**Review by mshinwell — APPROVED** [review]

**Review by xclerc — APPROVED** [review]
(Only for "backend/asmgen.ml".)

**Review by poechsel — APPROVED** [review]

File: driver/flambda_backend_args.mli (hunk 1 above)
Comment (gretay-js) [reply driver/flambda_backend_args.mli]: done

**Pull request merged**